src/grpc/infer_handler.h

```diff
@@ -652,12 +651,137 @@ class InferHandlerState {
+    {
+      std::lock_guard<std::mutex> lock(mu_);
+      inflight_states_.erase(state);
+    }
+
+    // Issues the cancellation for all inflight requests
+    // being tracked by this context.
+    TRITONSERVER_Error* IssueRequestCancellation()
+    {
+      TRITONSERVER_Error* err = nullptr;
+      {
+        std::lock_guard<std::mutex> lock(mu_);
+
+        // Issues the request cancellation to the core.
+        for (auto state : inflight_states_) {
+          LOG_VERBOSE(1) << "Issuing cancellation for " << state->unique_id_;
+          bool is_cancelled = false;
+          err = TRITONSERVER_InferenceRequestIsCancelled(
+              state->irequest_ptr_, &is_cancelled);
+          if ((err == nullptr) && (!is_cancelled)) {
+            err = TRITONSERVER_InferenceRequestCancel(state->irequest_ptr_);
+          }
+          if (err != nullptr) {
+            const char* request_id = "";
+            TRITONSERVER_Error* tmp_err = TRITONSERVER_InferenceRequestId(
+                state->irequest_ptr_, &request_id);
+            if (tmp_err != nullptr) {
+              // Ignore if failed to get request id.
+              TRITONSERVER_ErrorDelete(tmp_err);
+            }
+            TRITONSERVER_Error* wrapped_err = TRITONSERVER_ErrorNew(
+                TRITONSERVER_ERROR_INTERNAL,
+                std::string(
+                    std::string("[request id:") + request_id +
+                    "]: " + TRITONSERVER_ErrorMessage(err))
+                    .c_str());
+            TRITONSERVER_ErrorDelete(err);
+            return wrapped_err;
+          }
+
+          {
+            std::lock_guard<std::recursive_mutex> lock(state->step_mtx_);
+            if (state->step_ != Steps::CANCELLED) {
+              state->step_ = Steps::CANCELLATION_ISSUED;
+            }
+          }
+        }
+      }
+      return nullptr;
+    }
+
+
+    // Handles the gRPC context cancellation. This function can be called
+    // multiple times and is supposed to be re-entrant.
+    // Returns whether or not to continue cycling through the gRPC
+    // completion queue or not.
+    bool HandleCancellation(
+        InferHandlerStateType* state, bool rpc_ok, const std::string& name)
+    {
+      if (!IsCancelled()) {
+        LOG_ERROR << "[INTERNAL] HandleCancellation even when the context was "
+                     "not cancelled for "
+                  << name << ", rpc_ok=" << rpc_ok << ", context "
+                  << state->context_->unique_id_ << ", " << state->unique_id_
+                  << " step " << state->step_;
+        return true;
+      }
+      if ((state->step_ != Steps::CANCELLATION_ISSUED) &&
+          (state->step_ != Steps::CANCELLED)) {
+        LOG_VERBOSE(1) << "Cancellation notification received for " << name
+                       << ", rpc_ok=" << rpc_ok << ", context "
+                       << state->context_->unique_id_ << ", "
+                       << state->unique_id_ << " step " << state->step_;
```

Comment:
Side note, not for this PR, but it might be nice to have some lightly wrapped object or helper around some of these values used for logging purposes:
```cpp
struct RpcCallInfo {
  State state;
  bool rpc_ok;
};

ostream operator ...
{
  out <<  ", rpc_ok=" << rpc_ok << ", context "
        << state->context_->unique_id_ << ", "
        << state->unique_id_ << " step " << state->step_;
  return out;
}
```
So when we call this logging in most of our methods, it can be simplified to something like:
```cpp
LOG_VERBOSE(1) << "Cancellation notification received for: " << RpcCallInfo(state, rpc_ok);
```
Or something along those lines.